### ISSUE TITLE
[3.0] Remove dependence from Symfony Http-Foundation's Request

### ISSUE THREAD

**Issue by a3gz** [cartalyst/sentinel#181]
Hello. 

I have found that Symfony's Request, working along Slim framework v3 may cause problems. Specifically, Slim will loose access to the `php://input` stream under certain circumstances.

Given that the only use Sentinel seems to make of Symfony's request object is to get the client's IP; it would seem that removing said dependency would render Sentinel more independent with relatively little cost.

Best regards.

**Comment by brunogaspar** [comment]
> Specifically, Slim will loose access to the php://input stream under certain circumstances.

What are these `circumstances` you're talking about?

**Comment by a3gz** [comment]
If we:

*  Setup Sentinel with native integration. 
* Interact with Sentinel via an object obtained like this: `Sentinel::instance()->getSentinel()` before the Slim application is ran. 

The `php://input` stream will be empty, because `symfony/http-foundation` would have consumed already. 

I realize that this specific scenario is a problem only if we grab the Sentinel object before `$slim->run()`. If we grabbed it from within the route's closure, everything would be fine. 

I don't know how dependent will Sentinel be from that particular Symfony component as new features are added; but there is a source of conflict in the context of a native integration since read-once resources are consumed more than once.

Best regards. 


PS: 

Here is a Slim program that reproduces the circumstances I mentioned: 

```php 
    include "vendor/autoload.php";

    use Cartalyst\Sentinel\Native\Facades\Sentinel;

    $slim = new \Slim\App( );

    //
    // Here I grab the Sentinel object before $slim->run()
    // This is the problematic circumstance.
    $sentinel = Sentinel::instance()->getSentinel();

    $slim->put('/', function($request, $response) {
        // If we grab the object here, everything works, because Slim captured the stream
        // before Symfony.
        // $sentinel = Sentinel::instance()->getSentinel();

        //
        // If we grabbed the Sentinel object before $slim->run() the body will be empty.
        $data = $request->getBody()->getContents();
        $r = json_encode([
            'data' => $data
        ]);
        
        $body = $response->getBody();
        $body->write( $r );
        return $response->withBody($body);
    });

    $slim->run();
```

**Comment by brunogaspar** [comment]
So, you're saying that any package that makes use of Symfony Http Foundation and is used on Slim has this issue?

If that's the case, then it's an issue on either Slim or Symfony and it needs to be addressed there.

But either way, as a temp fix, you can easily extend the Sentinel Native facade and apply a patch for your use case i believe, and change the ip detection to use `$_SERVER['REMOTE_ADDR']` instead, but IMO this is not a really good solution.

**Comment by a3gz** [comment]
Well, I'm not going that far, in fact; the conflict only occurs when the Symfony Request object processes the HTTP request before anyone else; because it reads the input stream making it unavailable for other consumers. And this only happens when we request the Sentinel instance.   
So it is a very specific scenario limited only to the Request object and with events occurring in a specific order. 

Perhaps this is one of those things that are so unlikely to happen that don't deserve a second thought and I was to quick in posting the proposal. At first glance however, requiring a whole package to solve such a specific task didn't seem entirely necessary. 

So it came to me the question about whether native Sentinel integrations would become more dependent from Symfony's Request object, increasing opportunities for conflicts. 
As I think more about this, I realize that such eventuality would only be a problem if Sentinel actually needed, for some reason, to read the input stream... which may never happen.


As for solutions to my specific scenario: I first fixed using the same mechanism you suggested (thank you very much for it by the way); then I found another way using a closure to delay the creation of the Sentinel instance so it occurs after Slim runs.
